src/storage/src/source/reclock.rs

```diff
@@ -30,10 +30,193 @@ use mz_persist_client::read::{Listen, ListenEvent, ReadHandle};
 use mz_persist_client::write::WriteHandle;
 use mz_persist_client::Upper;
 use mz_repr::Timestamp;
+use tracing::trace;
 
 use crate::controller::CollectionMetadata;
 use crate::types::sources::MzOffset;
 
+/// TODO
+pub struct ReclockFollower {
+    /// A dTVC trace of the remap collection containing all consolidated updates at
+    /// `t` such that `since <= t < upper` indexed by partition and sorted by time.
+    remap_trace: HashMap<PartitionId, Vec<(Timestamp, MzOffset)>>,
+    /// Since frontier of the partial remap trace
+    since: Antichain<Timestamp>,
+    /// Upper frontier of the partial remap trace
+    upper: Antichain<Timestamp>,
+    /// The upper frontier in terms of `SourceTime`. Any attempt to reclock messages beyond this
+    /// frontier will lead to minting new bindings.
+    pub source_upper: HashMap<PartitionId, MzOffset>,
+}
+
+impl ReclockFollower {
+    /// Construct a new [ReclockOperator] from the given collection metadata
+    pub fn new(as_of: Antichain<Timestamp>) -> Self {
+        Self {
+            remap_trace: HashMap::new(),
+            since: as_of,
+            upper: Antichain::from_elem(Timestamp::minimum()),
+            source_upper: HashMap::new(),
+        }
+    }
+
+    /// Pushes new trace updates into this [`ReclockFollower`].
+    pub fn push_trace_updates(
+        &mut self,
+        updates: impl Iterator<Item = (PartitionId, Vec<(u64, MzOffset)>)>,
```

Comment:
The trait expected by`for` loops is `IntoIterator` so we should change that to `impl IntoIterator` to make life easier for callers.

Having the type like that will allow a caller to pass a `HashMap<PartitionId, Vec<(u64, MzOffset)>>` directly so we can get rid of all the `trace_updates.into_iter().collect()`calls below

Reply:
:+1: